### PR TITLE
feature: Add mpi support for mxnet estimator api

### PR DESCRIPTION
**Pull request by ChaiBapchya**
*Issue #, if available:* None

*Description of changes:*
Mxnet supports horovod for distributed training.
However, sagemaker-mxnet python sdk doesn't expose the horovod support.
This PR includes changes to allow end-users to invoke horovod [via mpi] as a key:value parameter in "distributions" dictionary

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to any/all clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ChaiBapchya** [comment]
sagemaker-python-sdk-notebook-tests fails at
```
FAILED	./amazon-sagemaker-examples/sagemaker-python-sdk/scikit_learn_randomforest/Sklearn_on_SageMaker_end2end.ipynb	PT10M14.717S
```

**Comment by ChaiBapchya** [comment]
Not sure why unit-tests is failing
I tested locally for black-check it passed
```
black tests/integ/test_horovod_mx.py 
All done! ✨ 🍰 ✨
1 file left unchanged.
```

@ChuyangDeng any idea?

**Comment by laurenyu** [comment]
I checked out this PR locally, and ran `tox -e black-format`:

```
black-format run-test-pre: PYTHONHASHSEED='2081812688'
black-format run-test: commands[0] | black -l 100 ./
reformatted /Users/laurenyu/Documents/workspace/sagemaker-python-sdk/tests/integ/test_horovod_mx.py
All done! ✨ 🍰 ✨
1 file reformatted, 287 files left unchanged.
```

and here are the changes if you're curious:

```
$ git diff
diff --git a/tests/integ/test_horovod_mx.py b/tests/integ/test_horovod_mx.py
index 58dde6ee..35121b5a 100644
--- a/tests/integ/test_horovod_mx.py
+++ b/tests/integ/test_horovod_mx.py
@@ -35,8 +35,7 @@ def test_hvd_cpu(sagemaker_session, cpu_instance_type, tmpdir):

 @pytest.mark.canary_quick
 @pytest.mark.skipif(
-    integ.test_region() in integ.TRAINING_NO_P2_REGIONS,
-    reason="no ml.p2 instances in this region",
+    integ.test_region() in integ.TRAINING_NO_P2_REGIONS, reason="no ml.p2 instances in this region"
 )
 def test_hvd_gpu(sagemaker_session, gpu_instance_type, tmpdir):
     _create_and_fit_estimator(sagemaker_session, gpu_instance_type, tmpdir)
```

if you want to use black directly instead of tox, you need to do `black -l 100 tests/integ/test_horovod_mx.py`

**Comment by ChaiBapchya** [comment]
@laurenyu Thanks a lot for the explanation. I ignored that `-l 100` by mistake.

**Comment by laurenyu** [comment]
re: latest failure - my guess is that this is due to different versions of black. if you look at the difference between [your change](https://github.com/aws/sagemaker-python-sdk/pull/1581/commits/3d50c465051a9f227aa5308e477a2dbfca10f903) and the diff I posted above, yours has a trailing comma at the end of the changed line.

File: src/sagemaker/fw_utils.py [hunk 1]
```diff
@@ -613,7 +613,7 @@ def warn_if_parameter_server_with_multi_gpu(training_instance_type, distribution
 
 
     """
-    if training_instance_type == "local" or distributions is None:
+    if training_instance_type == "local" or distributions is None or "mpi" in distributions:
```
Comment (laurenyu) [inline src/sagemaker/fw_utils.py:624]: I don't think you need to add mpi here - see ll. 624-626 with `ps_enabled`

File: src/sagemaker/mxnet/estimator.py [hunk 2]
```diff
@@ -99,8 +99,43 @@ def __init__(
                     * ``custom-image:latest``
 
             distributions (dict): A dictionary with information on how to run distributed
-                training (default: None). To have parameter servers launched for training,
-                set this value to be ``{'parameter_server': {'enabled': True}}``.
+                training (default: None). Currently we support distributed training with
+                parameter server and MPI [Horovod].
+                To enable parameter server use the following setup:
```
Comment (laurenyu) [inline src/sagemaker/mxnet/estimator.py:104]:
```suggestion
                parameter servers and MPI [Horovod].
                To enable parameter servers, use the following setup:
```

File: src/sagemaker/mxnet/estimator.py [hunk 3]
```diff
@@ -99,8 +99,43 @@ def __init__(
...
+                .. code:: python
+
+                    {
+                        'mpi':
+                        {
+                            'enabled': True
+                        }
+                    }
+
+                Option parameters within `mpi` are `processes_per_host` and `custom_mpi_options`.
```
Comment (laurenyu) [inline src/sagemaker/mxnet/estimator.py:126]:
```suggestion
                Option parameters within ``mpi`` are ``processes_per_host`` and ``custom_mpi_options``.
```

File: src/sagemaker/mxnet/estimator.py [hunk 4]
```diff
@@ -99,8 +99,43 @@ def __init__(
...
-                set this value to be ``{'parameter_server': {'enabled': True}}``.
+                training (default: None). Currently we support distributed training with
+                parameter server and MPI [Horovod].
+                To enable parameter server use the following setup:
+
+                .. code:: python
+
+                    {
+                        'parameter_server':
+                        {
```
Comment (laurenyu) [inline src/sagemaker/mxnet/estimator.py:110]: (same for the other two)

```suggestion
                        'parameter_server': {
```

File: tests/integ/test_horovod_mx.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import json
+import os
+import tarfile
+from six.moves.urllib.parse import urlparse
+
+import boto3
+import pytest
```
Comment (laurenyu) [inline tests/integ/test_horovod_mx.py:21]: (six is third-party)

```suggestion

import boto3
import pytest
from six.moves.urllib.parse import urlparse
```

File: tests/integ/test_horovod_mx.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+@pytest.mark.skipif(
+    integ.test_region() in integ.TRAINING_NO_P2_REGIONS, reason="no ml.p2 instances in this region"
+)
+def test_hvd_gpu(sagemaker_session, gpu_instance_type, tmpdir):
+    _create_and_fit_estimator(sagemaker_session, gpu_instance_type, tmpdir)
+
+
+@pytest.mark.local_mode
+@pytest.mark.parametrize("instances, processes", [[1, 2], (2, 1), (2, 2)])
+def test_horovod_local_mode(sagemaker_local_session, instances, processes, tmpdir):
```
Comment (laurenyu) [inline tests/integ/test_horovod_mx.py:46]: other than reading the output file, it doesn't seem like this test is providing much value for testing Python SDK functionality. would it be possible to instead put the assert at the end of the CPU test above?

File: src/sagemaker/mxnet/estimator.py (hunk 2 above)
Comment (ChaiBapchya) [reply src/sagemaker/mxnet/estimator.py:104]: I don't think there is a plural form for the term "parameter-server". Just like "Horovod", It's a Proper Noun not a common noun.

File: tests/integ/test_horovod_mx.py (hunk 6 above)
Comment (ChaiBapchya) [reply tests/integ/test_horovod_mx.py:46]: I'm following the same approach as horovod test for tensorflow in tests/integ/test_horovod.py
So for consistency purposes I'm sticking to a similar approach.

File: src/sagemaker/fw_utils.py (hunk 1 above)
Comment (ChaiBapchya) [reply src/sagemaker/fw_utils.py:624]: this itself is ll 624 right?

Comment (ChaiBapchya) [reply src/sagemaker/fw_utils.py:624]: Fair point
Got it!

File: src/sagemaker/mxnet/estimator.py (hunk 4 above)
Comment (ChaiBapchya) [reply src/sagemaker/mxnet/estimator.py:110]: Tensorflow estimator has same format
Why does the bracket have to be that way?
Screen Shot 2020-07-22 at 12 17 55 PM

File: tests/integ/test_horovod_mx.py (hunk 6 above)
Comment (laurenyu) [reply tests/integ/test_horovod_mx.py:46]: I could see both the TF and this one being removed. testing for the processes being set up correctly, etc. should live in the framework toolkit repositories, and here we just need to make sure that the hyperparameters get passed along correctly.

**Pull request merged**